### PR TITLE
std.tar: support pax headers and gnu_long{name,link}

### PR DESCRIPTION
**Pull request by travisstaloch**
tar.zig:
* add HeaderIterator() type and convert pipeToFileSystem() to use it.
* add tests parseNumeric and parsePaxTime ported from https://go.dev/src/archive/tar/strconv_test.go.
* add initial support for options.executable_bit_only.
* initial windows support:
  * skip symlinks which require admin rights
  * workaround file.updateTimes() panic by truncating file times

lib/std/compress/tar/testdata/
* copy a subset of tar files from https://go.dev/src/archive/tar/testdata

reader_test.zig:
* validate headers against files from testdata/. a port of https://go.dev/src/archive/tar/reader_test.go.

test_decompress.zig:
* runs tar.pipeToFileSystem() on valid testdata/ files.

### CODE REVIEW

**Comment by travisstaloch** [comment]
addresses, maybe closes #14310 
closes #15342 #15228
supersedes #15228

File: lib/std/tar.zig [hunk 1]
```diff
@@ -1,172 +1,1134 @@
...
+    // >= rather than > to make room for the null byte
+    if (file_path.len >= buf_len) return error.NameTooLong;
+    mem.copy(u8, buf, file_path);
+    buf[file_path.len] = 0;
+    return buf[0..file_path.len :0];
+}
+
+/// return the most significant, 'top' half of the time as an i64
+fn truncateTime(t: i128) i64 {
+    return @bitCast([2]i64, t)[0];
```
Comment (rohlem) [inline lib/std/tar.zig:1005]: I believe this is (unnecessarily) endian-specific; IMO the more idiomatic formulation is `@intCast(i64, t >> 64)`, which will work everywhere. (Shifting 1-bits out to the right using `>>` is currently allowed.)

Comment (travisstaloch) [reply lib/std/tar.zig]: Thanks! Fixed.

File: lib/std/compress/tar/test_decompress.zig [hunk 2]
```diff
@@ -0,0 +1,37 @@
+const std = @import("std");
+const mem = std.mem;
+const tar = std.tar;
+const testing = std.testing;
+
+test "tar decompress testdata" {
```
Comment (travisstaloch) [inline lib/std/compress/tar/test_decompress.zig:6]: The testdata/ folder is currently at 136K. I think each of the tar files should be gzipped.  And then this test, and maybe reader_test.zig, should use `@embedFile` rather than iterating the testdata/ directory.  Does that seem bettter? 

Thats whats being done in [the xz tests](https://github.com/ziglang/zig/blob/master/lib/std/compress/xz/test.zig#L21>

**Comment by travisstaloch** [comment]
I have updated the testdata/ files.  Previously they were .tar files and the testdata/ directory was around 136K.  Now they are now gzipped w/ -9 and total 6.9K (around 1/20 the size).  

Also, tests now use `@embedFile` rather than iterating the testdata/ direcotry.

File: lib/std/tar.zig [hunk 3]
```diff
@@ -1,172 +1,1141 @@
...
+
+    fn getFormat(h: V7Header) !FormatSet {
+        const value = try parseOctal(h.checksum());
+        const checksums = h.computeChecksum();
+
+        if (value != checksums[0] and value != checksums[1])
+            return fmt_unknown;
+
+        const magic = h.ustar().magic();
+        const version = h.ustar().version();
```
Comment (truemedian) [inline lib/std/tar.zig:561]: You never check the version for ustar headers. You could also simplify `magic` and `version` into the same field so that they get compared all at once. They're right next to each other and are always checked together. (ie. check ustar for `ustar  \x00` and gnu for `ustar\x0000`)

Also: such an 8 byte check could easily be optimized into a single 64 bit compare.

File: lib/std/tar.zig [hunk 4]
```diff
@@ -1,172 +1,1141 @@
...
+        return h.bytes[148..156];
+    }
+    /// Link indicator (file type) - 156..157
+    inline fn fileType(h: V7Header) FileType {
+        return @intToEnum(FileType, h.bytes[156]);
+    }
+    /// Name of linked file - 157..257
+    inline fn linkedFileName(h: V7Header) []const u8 {
+        return h.bytes[157..257];
     }
```
Comment (truemedian) [inline lib/std/tar.zig:516]: Why do these functions poke at bytes? These can be extern structs of byte arrays. Same for `UstarHeader`, `StarHeader` and `GnuHeader`

File: lib/std/tar.zig [hunk 5]
```diff
@@ -1,172 +1,1141 @@
...
+                        const amt = try self.reader.read(self.name_buf[0..self.buf.len]);
+                        if (amt != self.buf.len) return error.UnexpectedEndOfStream;
+                        gnu_long_name = mem.sliceTo(self.name_buf[0..self.buf.len], 0);
+                    },
+                    .gnu_long_link => {
+                        format.setIntersection(fmt_gnu);
+                        const amt = try self.reader.read(self.linkname_buf[0..self.buf.len]);
+                        if (amt != self.buf.len) return error.UnexpectedEndOfStream;
+                        gnu_long_link = mem.sliceTo(self.linkname_buf[0..self.buf.len], 0);
+                    },
```
Comment (truemedian) [inline lib/std/tar.zig:751]: These are wrong, you need to check the length of the header for the size of the longname or linkname. They can easily be over 512 bytes. These headers use the "file content" of the header as value.

**Review by truemedian — CHANGES_REQUESTED** [review]

File: lib/std/tar.zig (hunk 3 above)
Comment (travisstaloch) [reply lib/std/tar.zig:561]: I agree about the 64 bit compare.  But not sure the ustar version needs to be checked here.  As is, the result is `fmt_ustar_pax` which is a bitflags of `.ustar | .pax`. This is used later in `HeaderIterator.header()` where both ustar and pax are treated the same.  

I know it looks suspicious but i think it checks out. This is also how its done [in the go implementation](https://go.dev/src/archive/tar/format.go#L186) from which portions of this implementation are ported.

File: lib/std/tar.zig (hunk 4 above)
Comment (travisstaloch) [reply lib/std/tar.zig:516]: Sounds good. Will do. I was previously using extern structs with byte arrays but switched away for some reason that i can't remember now.  I think I was fumbling with a bitcast and settled on this.

File: lib/std/tar.zig (hunk 5 above)
Comment (travisstaloch) [reply lib/std/tar.zig:751]: Thanks. Dang, need a allocator now :shrug:

File: lib/std/tar.zig (hunk 3 above)
Comment (travisstaloch) [reply lib/std/tar.zig:561]: Note the use of the word 'guess' in the comments above `getFormat()` in the previous link. There are similar allusions [here](https://go.dev/src/archive/tar/common.go#L201) and [here](https://go.dev/src/archive/tar/reader.go#L162) which seem to indicate that determining the format isn't straightforward, at least in this implementation.

Comment (travisstaloch) [reply lib/std/tar.zig]: all reviews have been addressed in most recent commit. i'll allow you to decide wether to 'resolve conversation'.

File: lib/std/tar.zig (hunk 5 above)
Comment (travisstaloch) [reply lib/std/tar.zig]: went with an internal arena / page_allocator within `pipeToFileSystem()`. wasn't sure wether that function should accept one from the user.

File: lib/std/tar.zig [hunk 6]
```diff
@@ -1,172 +1,1086 @@
...
+        /// only. This bit is copied to the group and other executable bits.
+        /// Other bits of the mode are left as the default when creating files.
+        executable_bit_only,
+    };
+};
+
+pub fn pipeToFileSystem(dir: fs.Dir, reader: anytype, options: Options) !void {
+    var format = FormatSet.initMany(&.{ .ustar, .pax, .gnu });
+    var buf: [block_len * 2]u8 = undefined;
+    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
```
Comment (truemedian) [inline lib/std/tar.zig:1027]: This is certainly not a good idea. Allocators exist so that dynamic allocations like this aren't hidden.

**Review by truemedian — CHANGES_REQUESTED** [review]
Everything else here looks good.

File: lib/std/tar.zig (hunk 6 above)
Comment (travisstaloch) [reply lib/std/tar.zig:1029]: how does adding a `allocator: ?mem.Allocator` field to `Options` sound? maybe it shouldn't be optional?

Comment (truemedian) [reply lib/std/tar.zig:1029]: Oh, just realized that you did the same thing with pax headers, they may be more than 512 bytes and will need to be allocated. Better off making it required.

Comment (travisstaloch) [reply lib/std/tar.zig]: ended up just adding an allocator param to pipeToFileSystem()

File: lib/std/tar.zig [hunk 7]
```diff
@@ -1,172 +1,1096 @@
...
+        .pax_buf = buf[block_len .. 2 * block_len],
+        .arena = arena,
+    };
+}
+
+pub fn HeaderIterator(comptime Reader: type) type {
+    return struct {
+        reader: Reader,
+        buf: Block = undefined,
+        pax_buf: Block = undefined,
```
Comment (truemedian) [inline lib/std/tar.zig:634]: This has the same issue as the longname and longlink, it needs to be allocated because it can be >512 bytes.

Comment (travisstaloch) [reply lib/std/tar.zig:634]: Woops! Thank you. :+1:

Comment (travisstaloch) [reply lib/std/tar.zig]: addressed just now in forced push

Comment (travisstaloch) [reply lib/std/tar.zig]: glad you highlighted this. `Block` is actually defined as `*[512]u8`. The `= undefined` default was a remnant and has been removed now.

**Comment by travisstaloch** [comment]
I'm out of ideas for setting directory properties.  I've been doing the following.  But it doesn't seem to work in some cases, including on windows. 

```zig
fn setDirProperties(dir: fs.Dir, header: Header, options: Options) !void {
    // this is an attempt to re-use setFileProperties() and seems to cause many ci failures
    return setFileProperties(fs.File{ .handle = dir.fd }, header, options);
}
```

Afaict, the issue is that `std.fs.Dir.openDirW()` doesn't include `FILE_WRITE_ATTRIBUTES` and has no way to add it.  A fix might involve adding a `mode: std.fs.File.OpenMode` param to `std.fs.Dir.OpenDirOptions`  `std.fs.Dir.openDirW()`.  But this is outside the scope of this pr.

So I'm going to just replace `setDirProperties()` with a TODO.

File: lib/std/tar.zig (hunk 7 above)
Comment (travisstaloch) [reply lib/std/tar.zig]: addressed. i just rebased on latest master. i'm going to mark conversations previous to this point as resolved.

**Comment by squeek502** [comment]
The dir metadata stuff is https://github.com/ziglang/zig/issues/12377

**Comment by travisstaloch** [comment]
> The dir metadata stuff is #12377

Ahhh. This is exactly what i was running into. Thanks for pointing this out!

**Comment by travisstaloch** [comment]
Wondering what i should do about [this](https://github.com/travisstaloch/zig/blob/tar-addtests/lib/std/tar.zig#L1096)?

TLDR: seeing failures locally related to dir.symLink() but only in ReleaseSmall.  my guess is mis-compilation but haven't been able to reproduce with a test yet.

**Comment by travisstaloch** [comment]
I was able to reproduce the symlink issue with the following test. Turns out it was happening in debug modes too. But the buffer contents were set to undefined {170, 170...} and producing a bogus yet legal filename,  `''$'\252'` in bash where 252 is octal for 0xaa.

```zig
// /tmp/tmp2.zig
test {
    testing.log_level = .debug;
    const filename = "/home/travis/Documents/Code/zig/zig/lib/std/compress/tar/testdata/pax.tar.gz";
    const f = try std.fs.cwd().openFile(filename, .{});
    defer f.close();
    var gz_decomp = try std.compress.gzip.decompress(talloc, f.reader());
    defer gz_decomp.deinit();
    const reader = gz_decomp.reader();
    var dir = try std.fs.cwd().openDir("/tmp/asdf", .{});
    try std.tar.pipeToFileSystem(talloc, dir, reader, .{});
}
```

pax.tar contents:
```console
$ tree /tmp/asdf/
/tmp/asdf/
└── a
    ├── 123456789101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899100
    └── b -> 123456789101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899100
```

so it contains 2 files `a/1234...` and `a/b` - a symlink to `a/1234...`

```console
$ rm -rf /tmp/asdf/* && zig test /tmp/tmp2.zig
...
[default] (debug): symlink header.name=a/b header.linkname=123456789101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899100
[default] (debug): sub_path=�:{ 170 } basename=�:{ 170 } header.name={ 170, 170, 170 }
```

you can see here that before [this line](https://github.com/travisstaloch/zig/blob/eb0cd275b39a17cbc1005287e6d40cd6f02b3a6a/lib/std/tar.zig#L1121) `header.name` is fine. and afterward, it is corrupted. somehow it becomes corrupt [here in makeOpenPath()](https://github.com/travisstaloch/zig/blob/eb0cd275b39a17cbc1005287e6d40cd6f02b3a6a/lib/std/tar.zig#L1019) is my guess.

**Comment by travisstaloch** [comment]
I've pushed some changes to workaround the corruption issue and updated comments. see [comments and code here](https://github.com/travisstaloch/zig/blob/b459d90cbf96181b6defeb3e1b47fac9ad2e2bed/lib/std/tar.zig#L1095)

**Comment by travisstaloch** [comment]
I've pushed some changes to workaround the corruption issue and updated comments. see [comments and code here](https://github.com/travisstaloch/zig/blob/b459d90cbf96181b6defeb3e1b47fac9ad2e2bed/lib/std/tar.zig#L1095)

I'd appreciate any info about what might be happening and if there is a way to avoid it and/or what should be done.

**Comment by travisstaloch** [comment]
below is an almost self-contained reproduction of the symlink corruption issue. the only dependency is [this tar.zig file](https://github.com/travisstaloch/zig/blob/b459d90cbf96181b6defeb3e1b47fac9ad2e2bed/lib/std/tar.zig)

running it produces:
```console
...
[default] (debug): symlink header.name=a/b header.linkname=123456789101112131415161718192021222324252627282930313233343536373839404142434445464748495051525354555657585960616263646566676869707172737475767778798081828384858687888990919293949596979899100
[default] (debug): header.name=a/b:{ 97, 47, 98 }
[default] (debug): header.name=���:{ 170, 170, 170 } sub_path=�:{ 170 } basename=�:{ 170 }

====== expected this output: =========
a/b␃

======== instead found this: =========
���␃

======================================
First difference occurs on line 1:
expected:
a/b
^ ('\x61')
found:
���
^ ('\xaa')
Test [1/1] test_0... FAIL (TestExpectedEqual)
...
```

```zig
const std = @import("std");
const tar = @import("tar.zig");
// const tar = std.tar;
const testing = std.testing;
const talloc = testing.allocator;
const mem = std.mem;
const fs = std.fs;
const block_len = tar.block_len;
const log = std.log;

pub fn pipeToFileSystem(
    allocator: mem.Allocator,
    dir: fs.Dir,
    reader: anytype,
    options: tar.Options,
) !void {
    _ = options;
    var buf: [block_len]u8 = undefined;
    var iter = tar.headerIterator(reader, &buf, allocator);
    defer iter.deinit();

    while (try iter.next()) |header| {
        log.debug("pipeToFileSystem() header.type={}", .{header.type});
        switch (header.type) {
            .directory, .global_extended_header, .extended_header => {},
            .normal, .normal2, .hard_link => {
                const read_amt = mem.alignForwardGeneric(u64, @intCast(u64, header.size), block_len);
                try reader.skipBytes(read_amt, .{});
            },
            .symbolic_link => {
                std.log.debug("symlink header.name={s} header.linkname={s}", .{ header.name, header.linkname });
                if (fs.path.dirname(header.name)) |sub_path| {
                    const basename = header.name[sub_path.len + 1 ..];

                    // header.name is fine here
                    std.log.debug("header.name={s}:{any}", .{ header.name, header.name });
                    var expected = try allocator.dupe(u8, header.name);
                    defer allocator.free(expected);
                    var subdir = try dir.makeOpenPath(sub_path, .{});
                    std.log.debug("header.name={s}:{any} sub_path={s}:{any} basename={s}:{any}", .{ header.name, header.name, sub_path, sub_path, basename, basename });
                    try testing.expectEqualStrings(expected, header.name);
                    defer subdir.close();
                } else {}
            },
            else => {
                log.err("unsupported type '{s}'", .{@tagName(header.type)});
                return error.TarUnexpectedFileType;
            },
        }
    }
}

test {
    testing.log_level = .debug;
    // these are the bytes from https://github.com/travisstaloch/zig/blob/tar-addtests/lib/std/compress/tar/testdata/pax.tar.gz
    // the tar file contains 2 files `a/1234...` and `a/b` - a symlink to `a/1234...`
    const pax_tar_gz_bytes = [_]u8{ 31, 139, 8, 8, 49, 239, 65, 100, 2, 3, 112, 97, 120, 46, 116, 97, 114, 0, 237, 151, 75, 110, 219, 48, 16, 134, 181, 206, 41, 116, 2, 103, 222, 143, 69, 246, 93, 246, 10, 74, 98, 32, 65, 155, 160, 176, 93, 32, 199, 239, 84, 45, 138, 192, 136, 237, 69, 27, 1, 81, 231, 219, 80, 224, 80, 156, 33, 200, 95, 63, 53, 93, 127, 158, 94, 62, 109, 167, 251, 237, 110, 191, 177, 8, 191, 70, 98, 81, 243, 72, 4, 68, 36, 100, 20, 84, 52, 116, 12, 76, 2, 66, 34, 98, 18, 82, 50, 114, 10, 74, 6, 70, 38, 102, 22, 86, 54, 118, 14, 78, 1, 65, 33, 97, 17, 81, 129, 194, 68, 134, 106, 208, 21, 94, 183, 51, 34, 54, 32, 1, 155, 161, 18, 84, 63, 145, 24, 14, 227, 203, 176, 0, 223, 247, 135, 105, 87, 165, 252, 237, 60, 191, 215, 242, 167, 253, 32, 16, 200, 248, 109, 58, 60, 220, 76, 239, 176, 241, 38, 46, 33, 169, 160, 181, 177, 202, 42, 170, 106, 234, 26, 154, 6, 134, 70, 198, 38, 166, 102, 230, 22, 150, 14, 142, 78, 206, 46, 174, 110, 238, 30, 158, 1, 129, 65, 193, 33, 161, 97, 225, 17, 145, 9, 89, 5, 101, 37, 203, 154, 40, 107, 80, 86, 205, 112, 197, 48, 62, 29, 30, 159, 182, 55, 200, 10, 36, 146, 73, 27, 32, 174, 92, 8, 241, 51, 58, 157, 141, 222, 157, 142, 14, 171, 101, 137, 109, 159, 245, 111, 167, 245, 15, 224, 71, 250, 175, 138, 100, 24, 97, 73, 253, 239, 31, 166, 231, 237, 153, 113, 151, 226, 31, 84, 255, 243, 178, 118, 43, 62, 224, 205, 5, 253, 31, 251, 255, 237, 63, 207, 113, 209, 255, 235, 249, 151, 254, 203, 8, 168, 250, 177, 62, 192, 208, 254, 191, 140, 255, 219, 248, 245, 241, 249, 203, 124, 7, 88, 217, 13, 192, 140, 9, 54, 21, 160, 122, 153, 244, 232, 6, 240, 70, 244, 238, 116, 116, 189, 250, 191, 125, 247, 28, 179, 195, 187, 159, 241, 255, 35, 253, 83, 253, 50, 148, 255, 211, 18, 231, 241, 63, 247, 255, 166, 105, 154, 166, 105, 154, 166, 105, 154, 166, 105, 154, 166, 105, 154, 166, 105, 154, 102, 93, 252, 0, 138, 24, 188, 249, 0, 40, 0, 0 };
    var fbs = std.io.fixedBufferStream(&pax_tar_gz_bytes);
    var gz_decomp = try std.compress.gzip.decompress(talloc, fbs.reader());
    defer gz_decomp.deinit();
    const reader = gz_decomp.reader();

    var dir = try std.fs.cwd().makeOpenPath("/tmp/asdf", .{});
    defer dir.close();
    try pipeToFileSystem(talloc, dir, reader, .{});
}

```

**Comment by travisstaloch** [comment]
Turns out the perceived 'stack corruption' was just returning a pointer to a local and getting very 'lucky' until calling a function with a large enough stack to clobber it.  Thanks @jacobly0 for spotting it! on discord :pray:

**Comment by motiejus** [comment]
A cursory look and a nitpick ...

> I have updated the testdata/ files. Previously they were .tar files and the testdata/ directory was around 136K. Now they are gzipped w/ -9 and total 6.9K (around 1/20 the size).
> 
> Also, tests now use `@embedFile` on fixed lists of .gz files rather than iterating the testdata/ directory and opening/reading files.


Is this still a problem with reduced file sizes?

```
$ du -hs --apparent-size *
7.6K    gz
75K     plain
$ du -hs *
104K    gz
128K    plain
```

(`gz` are files in your PR, `plain` are extracted).

Most file systems have a block size of 4KiB. So each file takes a multiplier of 4KiB:

```
$ ls -1hsS plain/ 
total 128K
 12K pax.tar
8.0K pax-global-records.tar
8.0K xattrs.tar
8.0K gnu-multi-hdrs.tar
8.0K pax-multi-hdrs.tar
4.0K v7.tar
<...>
```

So if extracted, it would take 6 sectors more on disk (24 KiB). Granted, less sectors when extracted. But I argue that what matters more is the size of the repository when compressed (i.e. `git pull`):

```
$ tar -czf plain.tar.gz plain
$ tar -czf gz.tar.gz gz
$ stat -c "%n: %s" *.gz
gz.tar.gz: 7206
plain.tar.gz: 4475
```

So when the tarballs are compressed as a single unit, the un-compressed tarballs save around 3 KiB compared to the gzipped ones.

I recommend revising the decision to pre-compress them. It is rarely a good idea -- for the same reason we stopped compressing `stage1/zig.wasm` in c51288f1f6be20be9f162c086b6c1ef9382b30cd.

**Comment by motiejus** [comment]
> I recommend revising the decision to pre-compress them. It is rarely a good idea -- for the same reason we stopped compressing `stage1/zig.wasm` in [c51288f](https://github.com/ziglang/zig/commit/c51288f1f6be20be9f162c086b6c1ef9382b30cd).

See https://github.com/ziglang/zig/pull/13821#issuecomment-1344859738

**Comment by travisstaloch** [comment]
> I recommend revising the decision to pre-compress them. It is rarely a good idea -- for the same reason we stopped compressing `stage1/zig.wasm` in [c51288f](https://github.com/ziglang/zig/commit/c51288f1f6be20be9f162c086b6c1ef9382b30cd).

I see.  Does it matter that, unlike zig.wasm, these files are likely to never change?

And just to be clear, you're saying it would be better to go back to using .tar files than .tar.gz files?

**Comment by motiejus** [comment]
> > I recommend revising the decision to pre-compress them. It is rarely a good idea -- for the same reason we stopped compressing `stage1/zig.wasm` in [c51288f](https://github.com/ziglang/zig/commit/c51288f1f6be20be9f162c086b6c1ef9382b30cd).
> 
> I see. Does it matter that, unlike zig.wasm, these files are likely to never change?

I wouldn't have said it's a nitpick if they would change; then it would be quite a bit more important :)

> And just to be clear, you're saying it would be better to go back to using .tar files than .tar.gz files?

That's how I would do it. In the end, it kind of doesn't matter, but at least we can say that compressing does not give a lot of benefit, if any.

**Comment by travisstaloch** [comment]
I've rebased on latest and squashed all my commits together.  Forgive me if my squash formatting is non-standard.

**Comment by travisstaloch** [comment]
I don't think the ci failure has anything to do with this PR.  Its a docgen error.

**Comment by travisstaloch** [comment]
Thanks for re-running @andrewrk :+1:. All passing now.

**Comment by squeek502** [comment]
*Finally* got around to [setting up a fuzzer for this](https://github.com/squeek502/zig-std-lib-fuzzing/blob/master/fuzzers/tar.zig). Here's some initial results:

[tar-fuzzing-crashes-20230501.zip](https://github.com/ziglang/zig/files/11371094/tar-fuzzing-crashes-20230501.zip)

Mostly index-of-out-bounds errors found so far, with a few other panics mixed in. Note that the current fuzzer implementation will only check for crashes, not correctness.

The crashes should be reproducible by running the inputs through [`testDecompressTarToTmp`](https://github.com/ziglang/zig/pull/15382/files#diff-1db255cc936c695e2a6de190e0f146dabab040cebcd0cce4499b6215cc646c48R46-R50) with default options.

**Comment by travisstaloch** [comment]
Thank you @squeek502 :pray:! I've created [simple test runner](https://zigbin.io/dc4667) and fixed all the crashes from those files.  About to push a commit w/ the fixes.

**Comment by squeek502** [comment]
Looking good, only 2 new crashes found so far:

[tar-fuzzing-crashes-20230502.zip](https://github.com/ziglang/zig/files/11378186/tar-fuzzing-crashes-20230502.zip)

Looks like the first has to do with a `NUL` byte within a filepath which is asserted to not be the case in `std.os.toPosixPath`. This is actually a trickier scenario to handle than it appears, as all `fs`-related things currently treat invalid path names as `unreachable` (see https://github.com/ziglang/zig/pull/14533#issuecomment-1416888193 for another scenario this affects, and [here](https://github.com/squeek502/resinator/blob/88067d25ec9518071664a56efe20a54e33f60795/src/compile.zig#L357-L360) for where I ran into this same thing while fuzzing one of my projects). As an example, if a `tar` contained a file with a path that contains `|` (illegal on Windows), then it'd hit an `.OBJECT_NAME_INVALID => unreachable` when trying to create the file. To protect against hitting `unreachable` on all platforms, we'll likely either need:

- A robust path validation function that works properly for each platform. This seems possible (?) but it's likely we'll have to end up chasing corner cases for a while (and I'm unsure how different filesystems affect this, are illegal pathnames tied to the OS or to the particular filesystem that may be independent of the OS?).
- To abandon the 'invalid path => `unreachable`' strategy throughout `fs`-related std lib and instead treat invalid paths as a possible error everywhere. This seems more sensible to me personally, but I'm unsure if there are any benefits to the `=> unreachable` strategy.

I'm unsure of how (or if) this should be tackled within the `tar` implementation.

**Comment by travisstaloch** [comment]
> I'm unsure of how (or if) this should be tackled within the `tar` implementation. Might make sense to leave it as a follow-up issue.

I'm fine with leaving it as a follow up issue and perhaps adding TODOs which link to follow up issues.  I'll leave the decision for you or another maintainer.

Would erroring if a file path contains `NUL` be correct on all platforms?  If so, this would be easy to implement.  There are similar `NUL` checks happening when [validating pax headers](https://github.com/travisstaloch/zig/blob/8f260b95ffcd96869fbaed2acc6dbe64185463bb/lib/std/tar.zig#L862).

**Comment by squeek502** [comment]
I think erroring on `NUL` would be fine for now, then just leave a TODO about other potentially invalid paths. I've been meaning to investigate and write a proper issue about this.

**Comment by travisstaloch** [comment]
> Looking good, only 2 new crashes found so far:

Oh, after skipping paths with `NUL`s, a second crash was revealed - again related to an untagged `header.type` value.  This one caused an assertion fail from std.bit_set (due to the value being too big for the bitset).

I'll add a check that `header.type` is a tagged there to prevent that crash.

**Comment by travisstaloch** [comment]
> I think erroring on `NUL` would be fine for now, then just leave a TODO about other potentially invalid paths. I've been meaning to investigate and write a proper issue about this.

Just pushed a commit addressing the 2 new fuzz crashes.

**Comment by travisstaloch** [comment]
failing ci error doesn't seem related to this PR: 
> zig test ReleaseSafe native: error: memory usage peaked at 3776020480 bytes, exceeding the declared upper bound of 3738102579
<https://github.com/ziglang/zig/actions/runs/4867461146/jobs/8680068411?pr=15382#step:3:349>

**Comment by squeek502** [comment]
One new crash:

[tar-fuzzing-crashes-20230502.2.zip](https://github.com/ziglang/zig/files/11380180/tar-fuzzing-crashes-20230502.2.zip)

Will leave it running overnight and let you know if anything else gets found.

**Comment by squeek502** [comment]
12 hours of fuzzing later, all of the crashes found are the same `attempt to cast negative value to unsigned integer` panic from my last comment. Just for completeness, here are the other reproductions of the same crash:

[tar-fuzzing-crashes-20230503.zip](https://github.com/ziglang/zig/files/11389643/tar-fuzzing-crashes-20230503.zip)

Once this one is fixed, I think we're good to go on the fuzzing end of things.

**Comment by travisstaloch** [comment]
I've just pushed a commit addressing the remaining fuzzing crashes (from 2.2 and 3).  Here is a slightly modified [test script](https://zigbin.io/78cee1) for running the fuzz cases.

**Comment by squeek502** [comment]
Looks good from my end. Fuzzer is running cleanly.

**Comment by travisstaloch** [comment]
Really appreciate the fuzzing work @squeek502! Its been super nice to have someone find and provide the adverse inputs.  Certainly made my work a lot easier :pray:. 

Not sure who I should ask, but I think this is ready for review @andrewrk and @Vexu if you get a chance.

**Comment by travisstaloch** [comment]
@andrewrk i've just rebased and added a commit which uses the new builtin names.  cc @alichraghi who asked me to do so on discord.

**Comment by travisstaloch** [comment]
I don't think the aarch64-macos-debug ci failure is related to this PR.  Seems to be an OOM.

From the log:
```sh
zig test ReleaseSafe native: error: memory usage peaked at 4572626944 bytes, exceeding the declared upper bound of 4080218931
```

**Comment by jacobly0** [comment]
Yeah you're fine, fix is on master if you want to rebase.

**Comment by slimsag** [comment]
@travisstaloch looks like you have some more merge conflicts - just thought I'd mention as I am very interested in this PR.

**Comment by travisstaloch** [comment]
> @travisstaloch looks like you have some more merge conflicts - just thought I'd mention as I am very interested in this PR.

thanks for the heads up. just rebased and pushed.

**Comment by andrewrk** [comment]
Thanks for keeping this up to date. I'm sorry for taking so long with the review but I want you to know that it's on my mind, definitely looking to get it landed before the release.

**Comment by travisstaloch** [comment]
Thanks for the update. No worries, I know there is a lot to do.

**Comment by slimsag** [comment]
Not helpful-I know, but wanted to mention I am checking this PR every single day for updates as it is the #1 most impactful thing for us in terms of being able to get rid of submodules in Mach. We have a handful of packages ready-to-go but we just can't depend on them yet due to std.tar not being able to extract the archive.

File: src/Package.zig [hunk 8]
```diff
@@ -600,7 +600,7 @@ fn unpackTarball(
     var decompress = try compression.decompress(gpa, br.reader());
     defer decompress.deinit();
 
-    try std.tar.pipeToFileSystem(out_dir, decompress.reader(), .{
+    try std.tar.pipeToFileSystem(gpa, out_dir, decompress.reader(), .{
```
[the remark this answers is not in the document]
Comment (travisstaloch) [reply src/Package.zig:603]: I agree. But if we want to support long file names, it is necessary to have an allocator.  See @truemedian's comment https://github.com/ziglang/zig/pull/15382#discussion_r1174261182

File: lib/std/tar.zig [hunk 9]
```diff
@@ -1,177 +1,1198 @@
...
+            var format = fmt_ustar_pax_gnu;
+            var hdr = Header{};
+            var pax_hdr = Header{};
+
+            while (true) {
+                const v7 = try self.nextV7Header() orelse return null;
+                hdr = try self.header(v7) orelse return null;
+
+                format.setIntersection(hdr.fmt);
+                log.debug("hdr={}", .{hdr});
```
[the remark this answers is not in the document]
Comment (travisstaloch) [reply lib/std/tar.zig:698]: Thanks. I'm not planning to leave these log statements in.  I just didn't remove them yet anticipating more reviews and avoiding ci churn.

File: src/Package.zig (hunk 8 above)
[the remark this answers is not in the document]
Comment (squeek502) [reply src/Package.zig:603]: @travisstaloch it might be possible to use a buffer of size `std.fs.MAX_NAME_BYTES` and read the path one-component-at-a-time (erroring if any component is > `MAX_NAME_BYTES` long since it won't be possible to create it on the filesystem), using the `Dir` of the last created component to create the next component and so on.

File: lib/std/tar.zig [hunk 10]
```diff
@@ -1,177 +1,1198 @@
...
+                        if (gnu_long_name.len > 0) hdr.name = gnu_long_name;
+                        if (gnu_long_link.len > 0) hdr.linkname = gnu_long_link;
+                        if (hdr.type == .normal2) {
+                            hdr.type = if (mem.endsWith(u8, hdr.name, "/"))
+                                .directory
+                            else
+                                .normal;
+                        }
+
+                        // Set the final guess at the format.
```
[the remark this answers is not in the document]
Comment (travisstaloch) [reply lib/std/tar.zig:743]: I don't understand all the details the tar format.  My understanding it that It is messy and attempts to support old extensions.  Maybe someone else can help explain why this guessing is necessary.

Here are some references.

https://www.gnu.org/software/tar/manual/html_node/Standard.html
https://go.dev/src/archive/tar/common.go#L201
https://go.dev/src/archive/tar/format.go#L169

**Comment by travisstaloch** [comment]
> it looks like it blindly ports Go logic rather than building a tar implementation on first principles

You're right.  It does blindly port Go logic.  As much as I might want to help, I don't think I'm capable of creating a tar implementation from first principles.  I'll step aside and let someone else work on this.  Feel free continue with this PR or use anything from it in a different PR.

File: src/Package.zig (hunk 8 above)
[the remark this answers is not in the document]
Comment (truemedian) [reply src/Package.zig:603]: `MAX_NAME_BYTES` may work, but it needs to be able to reasonably handle windows systems (where such value is 260). You can't use the directory of the last component because tar files aren't necessarily laid out as a tree, you always have to create the file from the extraction root.

[the remark this answers is not in the document]
Comment (squeek502) [reply src/Package.zig:603]: 260 is `MAX_PATH` on Windows (and is irrelevant to Zig since we use extended-length path APIs exclusively). See `std.os.NAME_MAX`:

https://github.com/ziglang/zig/blob/f5239677e27a25beb450d2361fb15987d56de053/lib/std/os/windows.zig#L3595-L3611

I'm not sure what you mean by the second part. If each file in a tar has a filepath from the root, then, for example:

```
some-long-component/some-long-component/some-long-component/some-long-component/some-long-component/foo.txt
```

I'm suggesting something like:

- Read into `[MAX_NAME_BYTES]u8` until either you hit a path separator or run out of room (in which case, fatal error)
- Use a temporary buffer of one byte to check if this is the last component (read into it until you hit a non-slash or end-of-filepath)
- If this is not the last component, create and open it as a directory (using either the `Dir` of the root or the last created component's `Dir`)
  + If this is the last component, create it as a file and move to the next
- If there are more components to read, put the byte from the temporary buffer into the `[MAX_NAME_BYTES]u8` buffer and repeat from the start

**Comment by andrewrk** [comment]
Thank you and I apologize for the harsh words. I know a lot of people are counting on this improvement

**Comment by travisstaloch** [comment]
No worries.  I support the desire for high quality code.  And understand the desire for smaller PRs.  I'll continue to do what I can to help with std.tar.  I just don't feel confident in my ability to create a from scratch implementation.

**Comment by andrewrk** [comment]
Understood. I've been working on the release notes and your name has come up quite a few times. This release is regardless going to be much better off because of you.
